[WIP] LFortran code printing

anpandey opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests . Please also
write a comment on that issue linking back to this pull request once it is
open. -->


#### Brief description of what is fixed or changed

This pull request adds functionality to convert a SymPy expression to an [LFortran](https://lfortran.org/) abstract syntax tree node. Fortran code may then be generated utilizing the LFortran library.

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* codegen
  * add support for code generation through LFortran project
<!-- END RELEASE NOTES -->

anpandey committed:
Add initial attempt at sympy expression to LFortran converter

anpandey committed:
Modify the ASR conversion class to extend CodePrinter

The ASR converter extends `CodePrinter`, and the conversion functions now follow
the naming conventions of methods in the `CodePrinter` class. This will allow us
to use some of the existing methods in the class in the future.

czgdp1807 commented on sympy/codegen/lfort.py:
```diff
@@ -0,0 +1,85 @@
+        convertmethod = '_print_' + type(expr).__name__
+        if hasattr(self, convertmethod):
+            return getattr(self, convertmethod)(expr, **kwargs)
+        else:
```
This `else` isn't needed.

asmeurer replied:
Is it actually necessary to define this method at all? It should be handled in the superclass.

anpandey replied:
It works with the superclass method, but opens up the possibility of the printer returning something other than an AST. I've gotten rid of it for now, since it seems this can be adjusted in the superclass settings.

sylee957 commented on sympy/codegen/lfort.py:
```diff
@@ -0,0 +1,85 @@
+from sympy.printing.codeprinter import CodePrinter
+
+import lfortran.asr.asr as asr
+import lfortran.asr.builder as builder
```
There is `import_module` in `sympy.external`, or the external imports may fail

anpandey committed:
Use `sympy.external` for importing lfortran

Import errors can then be handled by sympy.

anpandey committed:
Change up some of the functions

`sympy_to_lfortran` instead of `sympy_to_asr` to more accurately reflect the
output. Also split into to functions: one that wraps and one that doesn't.

ArighnaIITG commented on sympy/codegen/lfort.py:
```diff
@@ -0,0 +1,88 @@
+from sympy.external import import_module
+
+lfortran = import_module("lfortran")
+asr = lfortran.asr.asr
```
You might have to add a `None` check here.

anpandey committed:
Add more (untested) visitor methods

- Float
- Rational
- Pow
- Function

anpandey committed:
Add initial infrastructure for testing

anpandey committed:
Make the testing functions more convenient

anpandey committed:
Default symbols to real

The other code generators also assume that symbols are real numbers

anpandey committed:
Fix floating point conversions

Needed to initialize the superclass to access `_settings`

anpandey committed:
Add negative one printer

Currently cannot be tested, because the AST -> ASR node visitor hasn't been
implemented

anpandey committed:
Order the terms in a binary expression

Follows the same pattern as `fcode`, resulting in the same AST.

anpandey committed:
Fix infinite loop on printing Mult

`_as_ordered_terms` doesn't split Mult, so the recursive call was being passed
with the same arguments

anpandey committed:
Add some preliminary tests

anpandey committed:
Remove unneeded print method

czgdp1807 commented:
@anpandey Please update on this.

oscargus commented:
Closing and reopening to trigger the tests.
